pkg/ingester/shipper.go

```diff
@@ -0,0 +1,289 @@
+	meta.Uploaded = nil
+
+	var uploadErrs int
+
+	metas, err := s.blockMetasFromOldest()
+	if err != nil {
+		return 0, err
+	}
+	for _, m := range metas {
+		// Do not sync a block if we already uploaded or ignored it. If it's no longer found in the bucket,
+		// it was generally removed by the compaction process.
+		if _, uploaded := hasUploaded[m.ULID]; uploaded {
+			meta.Uploaded = append(meta.Uploaded, m.ULID)
+			continue
+		}
+
+		if m.Stats.NumSamples == 0 {
+			// Ignore empty blocks.
+			level.Debug(s.logger).Log("msg", "ignoring empty block", "block", m.ULID)
+			continue
+		}
+
+		// We only ship of the first compacted block level as normal flow.
+		if m.Compaction.Level > 1 {
+			continue
+		}
+
+		// Check against bucket if the meta file for this block exists.
+		ok, err := s.bucket.Exists(ctx, path.Join(m.ULID.String(), block.MetaFilename))
+		if err != nil {
+			return 0, errors.Wrap(err, "check exists")
+		}
+		if ok {
+			meta.Uploaded = append(meta.Uploaded, m.ULID)
+			continue
+		}
+
+		if err := s.upload(ctx, m); err != nil {
+			// No error returned, just log line. This is because we want other blocks to be uploaded even
+			// though this one failed. It will be retried on second Sync iteration.
+			level.Error(s.logger).Log("msg", "shipping failed", "block", m.ULID, "err", err)
+			uploadErrs++
+			continue
+		}
+		meta.Uploaded = append(meta.Uploaded, m.ULID)
+		uploaded++
+		s.metrics.uploads.Inc()
+	}
+	if err := shipper.WriteMetaFile(s.logger, s.dir, meta); err != nil {
+		level.Warn(s.logger).Log("msg", "updating meta file failed", "err", err)
+	}
+
+	s.metrics.dirSyncs.Inc()
+	if uploadErrs > 0 {
+		s.metrics.uploadFailures.Add(float64(uploadErrs))
+		return uploaded, errors.Errorf("failed to sync %v blocks", uploadErrs)
+	}
+
+	return uploaded, nil
+}
+
+// sync uploads the block if not exists in remote storage.
+// TODO(khyatisoneji): Double check if block does not have deletion-mark.json for some reason, otherwise log it or return error.
+func (s *ThanosShipper) upload(ctx context.Context, meta *metadata.Meta) error {
+	level.Info(s.logger).Log("msg", "upload new block", "id", meta.ULID)
+
+	// We hard-link the files into a temporary upload directory so we are not affected
+	// by other operations happening against the TSDB directory.
+	updir := filepath.Join(s.dir, "thanos", "upload", meta.ULID.String())
+
+	// Remove updir just in case.
+	if err := os.RemoveAll(updir); err != nil {
+		return errors.Wrap(err, "clean upload directory")
+	}
+	if err := os.MkdirAll(updir, 0750); err != nil {
+		return errors.Wrap(err, "create upload dir")
+	}
+	defer func() {
+		if err := os.RemoveAll(updir); err != nil {
+			level.Error(s.logger).Log("msg", "failed to clean upload directory", "err", err)
+		}
+	}()
+
+	dir := filepath.Join(s.dir, meta.ULID.String())
+	if err := hardlinkBlock(dir, updir); err != nil {
+		return errors.Wrap(err, "hard link block")
+	}
```

Comment:
I would like to remove code that does "hardlinking" of block before uploading as well. WDYT? See also `hardlinkBlock` function.

Reply:
An user had issues because of this hard link:
https://github.com/grafana/mimir/discussions/1901

I think we can safely remove the hardlink. I think it was useful only in the Thanos sidecar context. That being said, I would rather do it in a separate PR, to keep it hidden here (given we squad commits, it's easier to keep track of it if we do it in a separate PR).

Reply:
>  That being said, I would rather do it in a separate PR

I agree, I would avoid blowing up this PR and make as many changes as possible in separate PRs to keep reviewing as easy as possible.